### PR TITLE
plugin: Add IDL definining the core API

### PR DESCRIPTION
**Pull request by abhinav**
This defines the underlying Thrift API that ThriftRW will use to talk to
plugins. Plugin authors will be provided a nicer API on top of this.

@prashantv @kriskowal @breerly

### CODE REVIEW

**Comment by abhinav** [comment]
CC @anuptalwalkar

**Comment by kriskowal** [comment]
Please include documentation for how to generate the code from the IDL.

:ok:

**Comment by abhinav** [comment]
@kriskowal I'll have another PR with the code generator for the plugin system and include that information there.

File: plugin/api.thrift [hunk 1]
```diff
@@ -0,0 +1,321 @@
...
+    /**
+     * Name of the plugin. This MUST match the name of the plugin specified
+     * over the command line or the program will fail.
+     */
+    1: required string name
+    /**
+     * Version of the plugin API.
+     *
+     * This is NOT the version of the plugin but the version of the plugin
+     * APIw hich the plugin respects. This MUST match the version that
```
Comment (prashantv) [inline plugin/api.thrift]: typo: `APIw hich` -> `API which`

File: plugin/api.thrift [hunk 2]
```diff
@@ -0,0 +1,321 @@
+/**
+ * Package api defines the core interface which plugins use to talk to
+ * ThriftRw.
```
Comment (prashantv) [inline plugin/api.thrift]: i saw `ThriftRW` in other places, also capitalize the `w` here?

File: plugin/api.thrift [hunk 3]
```diff
@@ -0,0 +1,321 @@
...
+/**
+ * ModuleID is an arbitrary unique identifier to reference the different
+ * modules in this request.
+ */
+typedef i32 ModuleID
+
+/**
+ * TypeReference is a reference to a user-defined type.
+ */
+struct TypeReference {
```
Comment (prashantv) [inline plugin/api.thrift]: Do type references need more information about whether the type they point to is a primitive or not?

Comment (prashantv) [inline plugin/api.thrift]: Never mind, looks like the `PointerType` will always wrap any pointers?

Comment (abhinav) [inline plugin/api.thrift]: Yep. Pointers will always be wrapped in that.

Comment (abhinav) [inline plugin/api.thrift]: See also https://github.com/thriftrw/thriftrw-go/pull/186/files#diff-12cd218d54cb1909fa3f8678286214e7R127

**Comment by prashantv** [comment]
lgtm

File: plugin/api.thrift (hunk 1 above)
Comment (abhinav) [inline plugin/api.thrift]: 👍

File: plugin/api.thrift (hunk 2 above)
Comment (abhinav) [inline plugin/api.thrift]: 👍

File: plugin/api.thrift [hunk 4]
```diff
@@ -0,0 +1,321 @@
...
+
+/**
+ * Version is the plugin API version specified in this IDL. This MUST be
+ * provided in HandshakeResponses to specify which version of the API the
+ * plugin implements.
+ */
+const i32 VERSION = 1
+
+/**
+ * ServiceID is an arbitrary unique identifier to reference the diffeerent
```
Comment (breerly) [inline plugin/api.thrift]: different*

File: plugin/doc.go [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+// Package plugin provides the API for writing ThriftRW plugins.
+//
+// Plugins are standalone programs with names in the format
+// thriftrw-plugin-$name where $name is the name of the plugin.
+//
+// 	// thriftrw-plugin-myfanceplugin/main.go
```
Comment (breerly) [inline plugin/doc.go]: myfancyplugin*

File: plugin/mock_service_generator_test.go [hunk 6]
```diff
@@ -0,0 +1,41 @@
+// Automatically generated by MockGen. DO NOT EDIT!
```
Comment (breerly) [inline plugin/mock_service_generator_test.go]: Wondering why mockgen doesn't produce a `// Code generated by mockgen` comment so theses would be collapsed.

Comment (breerly) [inline plugin/mock_service_generator_test.go]: https://github.com/golang/mock/issues/54

**Comment by breerly** [comment]
lgtm :shipit:

File: plugin/api.thrift (hunk 4 above)
Comment (abhinav) [inline plugin/api.thrift]: 👍

File: plugin/doc.go (hunk 5 above)
Comment (abhinav) [inline plugin/doc.go]: 👍

**Pull request merged**